backend/src/main/java/com/celuveat/restaurant/query/dao/RestaurantWithDistanceDao.java

```diff
@@ -138,6 +140,55 @@ private OrderSpecifier<?> applyOrderBy(Pageable pageable) {
         return distanceColumn.asc();
     }
 
+    public Page<RestaurantWithDistance> searchByAddress(
+            AddressSearchCond addressSearchCond,
+            Pageable pageable
+    ) {
+        List<RestaurantWithDistance> resultList = query.selectDistinct(Projections.constructor(
+                        RestaurantWithDistance.class,
+                        restaurant.id,
+                        restaurant.name,
+                        restaurant.category,
+                        restaurant.roadAddress,
+                        restaurant.latitude,
+                        restaurant.longitude,
+                        restaurant.phoneNumber,
+                        restaurant.naverMapUrl,
+                        restaurant.viewCount,
+                        numberTemplate(Double.class, "0"),
+                        restaurant.likeCount
+                ))
+                .from(restaurant)
+                .where(
+                        restaurantAddressIn(addressSearchCond.addresses)
+                ).orderBy(restaurant.likeCount.desc(), restaurant.id.asc())
+                .offset(pageable.getOffset())
+                .limit(pageable.getPageSize())
+                .fetch();
+
+        JPAQuery<Long> countQuery = query.select(restaurant.countDistinct())
+                .from(restaurant)
+                .where(
+                        restaurantAddressIn(addressSearchCond.addresses)
+                );
+        return PageableExecutionUtils.getPage(resultList, pageable, countQuery::fetchOne);
+    }
+
+    private BooleanExpression restaurantAddressIn(List<String> addresses) {
+        if (addresses.size() == 1) {
+            return restaurant.roadAddress.startsWith(addresses.get(0));
+        }
+
+        List<BooleanExpression> conditions = new ArrayList<>();
+        for (String address : addresses) {
+            conditions.add(restaurant.roadAddress.startsWith(address));
+        }
+        BooleanExpression orConditions = conditions.remove(0);
+        for (BooleanExpression condition : conditions) {
+            orConditions = orConditions.or(condition);
+        }
+        return orConditions;
```

Comment:
```suggestion
        return addresses.stream()
                .map(restaurant.roadAddress::startsWith)
                .reduce(BooleanExpression::or)
                .get();
```

이렇게 줄일 수 있을 것 같아.
위 코드로 말랑이 짠 테스트도 돌려봐서 동일한 동작하는지 확인해봤어!

Reply:
와우 너무 좋은데요?